crates/matrix-sdk-crypto/src/session_manager/sessions.rs

```diff
@@ -336,71 +379,84 @@ impl SessionManager {
             .failures
             .keys()
             .filter_map(|s| ServerName::parse(s).ok())
-            .filter(|s| s != self.store.static_account().user_id.server_name());
+            .filter(|s| s != self.store.static_account().user_id.server_name())
+            .collect();
         let successful_servers = response.one_time_keys.keys().map(|u| u.server_name());
 
-        self.failures.extend(failed_servers);
+        self.failures.extend(failed_servers.iter().cloned());
         self.failures.remove(successful_servers);
 
-        // build a map of user_id -> device_id -> key for each device we can start a
-        // session with...
-        let mut device_map: BTreeMap<
-            OwnedUserId,
-            BTreeMap<OwnedDeviceId, &Raw<ruma::encryption::OneTimeKey>>,
-        > = BTreeMap::new();
-
-        // ... and a list of (user_id, device_id) pairs where the one-time-key is
-        // missing
-        let mut missing_devices: Vec<(OwnedUserId, OwnedDeviceId)> = Vec::new();
-
-        for (user_id, user_devices) in response.one_time_keys.iter() {
-            for (device_id, key_map) in user_devices {
-                match key_map.values().next() {
-                    Some(k) => {
-                        device_map.entry(user_id.clone()).or_default().insert(device_id.clone(), k);
-                    }
-                    None => {
-                        missing_devices.push((user_id.clone(), device_id.clone()));
-                    }
-                };
-            }
-        }
+        // build a map of (user_id, device_id) -> key from the response
+        let sessions_to_create: BTreeMap<(OwnedUserId, OwnedDeviceId), _> = response
+            .one_time_keys
+            .iter()
+            .flat_map(|(user_id, device_map)| {
+                device_map.iter().filter_map(|(device_id, key_map)| {
+                    // For devices that returned at least one key, this will yield
+                    // `Some((user_id, device_id), first_key)`. For devices that returned an
+                    // empty map, this will yield `None`.
+                    //
+                    // The `None` entries are then filtered out by `filter_map`.
+                    key_map.values().next().map(|key| ((user_id.clone(), device_id.clone()), key))
+                })
+            })
+            .collect();
 
-        // process all the missing devices at once to save repeatedly grabbing the lock
-        if !missing_devices.is_empty() {
-            warn!(
-                ?missing_devices,
-                "Tried to create a new sessions, but the signed one-time key was missing for some devices",
-            );
+        // if we were able to pair this response with a request, look for devices that
+        // were present in the request but did not elicit a successful response.
+        if let Some(request) = request {
+            let missing_devices: Vec<(OwnedUserId, OwnedDeviceId)> = request
```

Comment:
ok, happy to do this. The reason I avoided it was that it felt a bit redundant to build the two sets just so that we could call `.difference` on them, particularly given we already have the `requested_sessions` in a map whose keys we can check in O(n log n) (ie, the same as the `BTreeSet`).

I guess using sets shows the intention more clearly though.

Reply:
Luckily we can check if the `BTreeSet` approach is slower, red is the `flat_map()`, blue is the `BTreeSet` approach:

![flat_map_vs_sets_page-0001](https://github.com/matrix-org/matrix-rust-sdk/assets/552026/3a67070e-8b4d-4dbc-ab18-e9d1b8c9eca8)

The benchmark is processing a `/keys/claim` response with 163 one-time keys. I did run it a couple of times to get somewhat stable results. `BTreeSet` seems to be either a bit better or at worst it makes no difference which approach we pick.

Reply:
To check I am understanding the report correctly: this is saying the mean time increased by 0.7680%? In which case, very much agreed this is not worth worrying about.

I'd be interested to see how you ran this benchmark, for future reference.

Reply:
You are understanding correctly, I did have a couple of runs which claimed that the time decreased by ~0.5% as well. As the report tells us, all of this is within the noise threshold.

> I'd be interested to see how you ran this benchmark, for future reference.

Please take a look at the `/benchmarks` folder:  https://github.com/matrix-org/matrix-rust-sdk/tree/main/benchmarks